File: frontend/src/constants/fleet-inventory.ts
/** KrishiFarms fleet inventory — mirrors scripts/data/fleet_inventory.py */

export interface FleetCatalogItem {
  code: string;
  name: string;
  nameTe: string;
  fuelType: "tractor" | "diesel" | "implement";
}

/** Ops dropdown types: Tractor … Drone */
export const FLEET_SERVICE_TYPES: FleetCatalogItem[] = [
  { code: "TRACTOR", name: "Tractor", nameTe: "ట్రాక్టర్", fuelType: "tractor" },
  { code: "CULTIVATOR", name: "Cultivator", nameTe: "కల్టివేటర్", fuelType: "implement" },
  { code: "ROTAVATOR", name: "Rotavator", nameTe: "రోటవేటర్", fuelType: "implement" },
  { code: "BALER", name: "Baler", nameTe: "బేలర్", fuelType: "implement" },
  { code: "TROLLEY", name: "Trolley", nameTe: "ట్రాలీ", fuelType: "implement" },
  { code: "WEEDER", name: "Weeder", nameTe: "వీడర్", fuelType: "implement" },
  {
    code: "FERTILIZER_PUMP",
    name: "Fertilizer Pump",
    nameTe: "ఎరువు పంపు",
    fuelType: "implement",
  },
  { code: "BOLERO", name: "Bolero", nameTe: "బోలేరో", fuelType: "diesel" },
  { code: "DCM", name: "DCM", nameTe: "డీసీఎం", fuelType: "diesel" },
  { code: "HARVESTER", name: "Harvester", nameTe: "హార్వెస్టర్", fuelType: "diesel" },
  { code: "DRONE", name: "Drone", nameTe: "డ్రోన్", fuelType: "implement" },
];

export const FLEET_TRACTORS: FleetCatalogItem[] = [
  {
    code: "JD_TRACTOR_2W",
    name: "John Deere tractor 2W",
    nameTe: "జాన్ డియర్ ట్రాక్టర్ 2W",
    fuelType: "tractor",
  },
  {
    code: "JD_TRACTOR_4W",
    name: "John Deere tractor 4W",
    nameTe: "జాన్ డియర్ ట్రాక్టర్ 4W",
    fuelType: "tractor",
  },
];

export const FLEET_TRANSPORT: FleetCatalogItem[] = [
  {
    code: "MAHINDRA_BOLERO",
    name: "Mahindra Bolero",
    nameTe: "మహీంద్రా బోలేరో",
    fuelType: "diesel",
  },
  {
    code: "EICHER_DCM",
    name: "Eicher DCM",
    nameTe: "ఐషర్ డీసీఎం",
    fuelType: "diesel",
  },
];

export const FLEET_IMPLEMENTS: FleetCatalogItem[] = FLEET_SERVICE_TYPES.filter((v) =>
  ["TROLLEY", "BALER", "FERTILIZER_PUMP", "CULTIVATOR", "ROTAVATOR", "WEEDER"].includes(v.code),
);

export const FLEET_LEGACY_CODES = ["PUMP"] as const;

export const FUEL_TYPE_OPTIONS = [
  { value: "tractor", label: "Tractor" },
  { value: "diesel", label: "Diesel vehicle" },
  { value: "implement", label: "Implement / attachment" },
] as const;

export const FLEET_INVENTORY_SUMMARY =
  "Tractor, Cultivator, Rotavator, Baler, Trolley, Weeder, Fertilizer Pump, Bolero, DCM, Harvester, Drone — plus John Deere 2W/4W, Mahindra Bolero, Eicher DCM";

export const TRACTOR_WORK_VEHICLE_CODES = new Set([
  "TRACTOR",
  "CULTIVATOR",
  "ROTAVATOR",
  "BALER",
  "TROLLEY",
  "WEEDER",
  "FERTILIZER_PUMP",
  "HARVESTER",
  "DRONE",
  "PUMP",
  ...FLEET_TRACTORS.map((v) => v.code),
]);

export const TRANSPORT_VEHICLE_CODES = new Set([
  "BOLERO",
  "DCM",
  ...FLEET_TRANSPORT.map((v) => v.code),
]);

export const FLEET_VEHICLE_ORDER = [
  ...FLEET_SERVICE_TYPES.map((v) => v.code),
  ...FLEET_TRACTORS.map((v) => v.code),
  ...FLEET_TRANSPORT.map((v) => v.code),
  ...FLEET_LEGACY_CODES,
];

/** Vehicle-specific work form profile (conditional questions). */
export type VehicleWorkProfile =
  | "tractor"
  | "trolley"
  | "bolero"
  | "dcm"
  | "pump"
  | "drone"
  | "baler"
  | null;

/** Farming implements share tractor-style crop/area/stage questions. */
const TRACTOR_PROFILE_CODES = new Set([
  "TRACTOR",
  "JD_TRACTOR_2W",
  "JD_TRACTOR_4W",
  "CULTIVATOR",
  "ROTAVATOR",
  "WEEDER",
  "HARVESTER",
]);
const BALER_PROFILE_CODES = new Set(["BALER"]);
const TROLLEY_PROFILE_CODES = new Set(["TROLLEY"]);
const BOLERO_PROFILE_CODES = new Set(["BOLERO", "MAHINDRA_BOLERO"]);
const DCM_PROFILE_CODES = new Set(["DCM", "EICHER_DCM"]);
const PUMP_PROFILE_CODES = new Set(["FERTILIZER_PUMP", "PUMP"]);
const DRONE_PROFILE_CODES = new Set(["DRONE"]);

export function resolveVehicleWorkProfile(
  code: string | undefined | null,
  fuelType?: string | null,
): VehicleWorkProfile {
  if (!code) return null;
  const c = code.toUpperCase();
  if (BALER_PROFILE_CODES.has(c)) return "baler";
  if (TRACTOR_PROFILE_CODES.has(c)) return "tractor";
  if (TROLLEY_PROFILE_CODES.has(c)) return "trolley";
  if (BOLERO_PROFILE_CODES.has(c)) return "bolero";
  if (DCM_PROFILE_CODES.has(c)) return "dcm";
  if (PUMP_PROFILE_CODES.has(c)) return "pump";
  if (DRONE_PROFILE_CODES.has(c)) return "drone";

  // Custom master-data codes (e.g. tractor4W) — infer from code / fuel_type.
  if (c.includes("TROLLEY")) return "trolley";
  if (c.includes("BOLERO")) return "bolero";
  if (c.includes("DCM")) return "dcm";
  if (c.includes("DRONE")) return "drone";
  if (c.includes("PUMP")) return "pump";
  if (c.includes("BALER")) return "baler";
  if (
    c.includes("TRACTOR") ||
    c.includes("CULTIVATOR") ||
    c.includes("ROTAVATOR") ||
    c.includes("WEEDER") ||
    c.includes("HARVESTER") ||
    /[24]W/.test(c)
  ) {
    return "tractor";
  }
  const fuel = (fuelType ?? "").toLowerCase();
  if (fuel === "tractor" || fuel === "implement") return "tractor";
  if (fuel === "diesel") {
    if (c.includes("BOLERO")) return "bolero";
    if (c.includes("DCM")) return "dcm";
  }
  return null;
}

interface VehicleLike {
  code: string;
  is_active?: boolean;
  fuel_type?: string | null;
}

function codeInSet(code: string, set: Set<string>): boolean {
  return set.has(code) || set.has(code.toUpperCase());
}

/** True when a custom vehicle type belongs in tractor / implement work lists. */
function isTractorWorkCustom(v: VehicleLike): boolean {
  const fuel = (v.fuel_type ?? "").toLowerCase();
  if (fuel === "tractor" || fuel === "implement") return true;
  if (fuel === "diesel") return false;
  const c = v.code.toUpperCase();
  return (
    c.includes("TRACTOR") ||
    c.includes("CULTIVATOR") ||
    c.includes("ROTAVATOR") ||
    c.includes("BALER") ||
    c.includes("WEEDER") ||
    c.includes("HARVESTER") ||
    c.includes("TROLLEY") ||
    c.includes("PUMP") ||
    c.includes("DRONE") ||
    /[24]W/.test(c)
  );
}

function isTransportCustom(v: VehicleLike): boolean {
  const fuel = (v.fuel_type ?? "").toLowerCase();
  if (fuel === "diesel") return true;
  if (fuel === "tractor" || fuel === "implement") return false;
  const c = v.code.toUpperCase();
  return c.includes("BOLERO") || c.includes("DCM");
}

export function filterVehiclesForCategory<T extends VehicleLike>(
  category: string,
  vehicles: T[],
): T[] {
  const active = vehicles.filter((v) => v.is_active !== false);
  if (category === "transport") {
    return active.filter(
      (v) => codeInSet(v.code, TRANSPORT_VEHICLE_CODES) || isTransportCustom(v),
    );
  }
  if (category === "tractor_work") {
    return active.filter(
      (v) => codeInSet(v.code, TRACTOR_WORK_VEHICLE_CODES) || isTractorWorkCustom(v),
    );
  }
  if (category === "vehicle_ops") {
    return active;
  }
  return active;
}

export function sortFleetVehicles<T extends VehicleLike>(vehicles: T[]): T[] {
  const order = new Map(FLEET_VEHICLE_ORDER.map((code, index) => [code, index]));
  return [...vehicles].sort((a, b) => {
    const ai = order.get(a.code) ?? 999;
    const bi = order.get(b.code) ?? 999;
    if (ai !== bi) return ai - bi;
    return a.code.localeCompare(b.code);
  });
}

File: scripts/data/fleet_inventory.py
"""KrishiFarms fleet inventory — service types, units, and implements.

Canonical catalog for field services, vehicle types master data, and fleet placeholders.
"""

from __future__ import annotations

# Dropdown / ops vehicle types (Tractor … Drone)
FLEET_SERVICE_TYPES: list[dict] = [
    {"code": "TRACTOR", "name": "Tractor", "name_te": "ట్రాక్టర్", "fuel_type": "tractor"},
    {"code": "CULTIVATOR", "name": "Cultivator", "name_te": "కల్టివేటర్", "fuel_type": "implement"},
    {"code": "ROTAVATOR", "name": "Rotavator", "name_te": "రోటవేటర్", "fuel_type": "implement"},
    {"code": "BALER", "name": "Baler", "name_te": "బేలర్", "fuel_type": "implement"},
    {"code": "TROLLEY", "name": "Trolley", "name_te": "ట్రాలీ", "fuel_type": "implement"},
    {"code": "WEEDER", "name": "Weeder", "name_te": "వీడర్", "fuel_type": "implement"},
    {
        "code": "FERTILIZER_PUMP",
        "name": "Fertilizer Pump",
        "name_te": "ఎరువు పంపు",
        "fuel_type": "implement",
    },
    {"code": "BOLERO", "name": "Bolero", "name_te": "బోలేరో", "fuel_type": "diesel"},
    {"code": "DCM", "name": "DCM", "name_te": "డీసీఎం", "fuel_type": "diesel"},
    {"code": "HARVESTER", "name": "Harvester", "name_te": "హార్వెస్టర్", "fuel_type": "diesel"},
    {"code": "DRONE", "name": "Drone", "name_te": "డ్రోన్", "fuel_type": "implement"},
]

# Named inventory units on the farm
FLEET_TRACTORS: list[dict] = [
    {
        "code": "JD_TRACTOR_2W",
        "name": "John Deere tractor 2W",
        "name_te": "జాన్ డియర్ ట్రాక్టర్ 2W",
        "fuel_type": "tractor",
    },
    {
        "code": "JD_TRACTOR_4W",
        "name": "John Deere tractor 4W",
        "name_te": "జాన్ డియర్ ట్రాక్టర్ 4W",
        "fuel_type": "tractor",
    },
]

FLEET_TRANSPORT: list[dict] = [
    {
        "code": "MAHINDRA_BOLERO",
        "name": "Mahindra Bolero",
        "name_te": "మహీంద్రా బోలేరో",
        "fuel_type": "diesel",
    },
    {
        "code": "EICHER_DCM",
        "name": "Eicher DCM",
        "name_te": "ఐషర్ డీసీఎం",
        "fuel_type": "diesel",
    },
]

# Default asset instances linked to fleet inventory vehicle types (seeded by seed_services).
DEFAULT_FLEET_ASSETS: list[dict] = [
    {
        "asset_code": "JD2W-01",
        "name": "John Deere tractor 2W",
        "name_te": "జాన్ డియర్ ట్రాక్టర్ 2W",
        "vehicle_type_code": "JD_TRACTOR_2W",
        "asset_category": "tractor",
        "fuel_type": "tractor",
    },
    {
        "asset_code": "JD4W-01",
        "name": "John Deere tractor 4W",
        "name_te": "జాన్ డియర్ ట్రాక్టర్ 4W",
        "vehicle_type_code": "JD_TRACTOR_4W",
        "asset_category": "tractor",
        "fuel_type": "tractor",
    },
    {
        "asset_code": "BOLERO-01",
        "name": "Mahindra Bolero",
        "name_te": "మహీంద్రా బోలేరో",
        "vehicle_type_code": "MAHINDRA_BOLERO",
        "asset_category": "bolero",
        "fuel_type": "diesel",
    },
    {
        "asset_code": "DCM-01",
        "name": "Eicher DCM",
        "name_te": "ఐషర్ డీసీఎం",
        "vehicle_type_code": "EICHER_DCM",
        "asset_category": "dcm",
        "fuel_type": "diesel",
    },
]

# Implements already covered by FLEET_SERVICE_TYPES (trolley, baler, pump, cultivator, rotavator, weeder)
FLEET_IMPLEMENTS: list[dict] = [
    row
    for row in FLEET_SERVICE_TYPES
    if row["code"]
    in {"TROLLEY", "BALER", "FERTILIZER_PUMP", "CULTIVATOR", "ROTAVATOR", "WEEDER"}
]

# Legacy codes kept for idempotent upsert on existing orgs (name refresh only).
LEGACY_VEHICLE_ALIASES: list[dict] = [
    {
        "code": "PUMP",
        "name": "Fertilizer Pump",
        "name_te": "ఎరువు పంపు",
        "fuel_type": "implement",
    },
]

DEFAULT_VEHICLE_TYPES: list[dict] = [
    *FLEET_SERVICE_TYPES,
    *FLEET_TRACTORS,
    *FLEET_TRANSPORT,
    *LEGACY_VEHICLE_ALIASES,
]

DEFAULT_ACTIVITY_TYPES: list[dict] = [
    # Field services
    {
        "code": "FIELD_PLOUGHING",
        "name": "Ploughing",
        "name_te": "దున్నుట",
        "service_category": "field_service",
        "default_rate_type": "hourly",
    },
    {
        "code": "FIELD_SPRAYING",
        "name": "Spraying",
        "name_te": "పిచికారీ",
        "service_category": "field_service",
        "default_rate_type": "hourly",
    },
    {
        "code": "FIELD_OTHER",
        "name": "Other Field Service",
        "name_te": "ఇతర ఫీల్డ్ సేవ",
        "service_category": "field_service",
        "default_rate_type": "fixed",
    },
    # Tractor work
    {
        "code": "TRACTOR_JD_2W",
        "name": "John Deere 2W Tractor Work",
        "name_te": "జాన్ డియర్ 2W ట్రాక్టర్ పని",
        "service_category": "tractor_work",
        "default_rate_type": "hourly",
    },
    {
        "code": "TRACTOR_JD_4W",
        "name": "John Deere 4W Tractor Work",
        "name_te": "జాన్ డియర్ 4W ట్రాక్టర్ పని",
        "service_category": "tractor_work",
        "default_rate_type": "hourly",
    },
    {
        "code": "TRACTOR_CULTIVATOR",
        "name": "Cultivator Work",
        "name_te": "కల్టివేటర్ పని",
        "service_category": "tractor_work",
        "default_rate_type": "hourly",
    },
    {
        "code": "TRACTOR_ROTAVATOR",
        "name": "Rotavator Work",
        "name_te": "రోటవేటర్ పని",
        "service_category": "tractor_work",
        "default_rate_type": "hourly",
    },
    {
        "code": "TRACTOR_TROLLEY",
        "name": "Trolley Work",
        "name_te": "ట్రాలీ పని",
        "service_category": "tractor_work",
        "default_rate_type": "fixed",
    },
    {
        "code": "TRACTOR_BALER",
        "name": "Baler Work",
        "name_te": "బేలర్ పని",
        "service_category": "tractor_work",
        "default_rate_type": "fixed",
    },
    {
        "code": "TRACTOR_PUMP",
        "name": "Fertilizer Pump Work",
        "name_te": "ఎరువు పంపు పని",
        "service_category": "tractor_work",
        "default_rate_type": "hourly",
    },
    {
        "code": "TRACTOR_WEEDER",
        "name": "Weeder Work",
        "name_te": "వీడర్ పని",
        "service_category": "tractor_work",
        "default_rate_type": "hourly",
    },
    {
        "code": "TRACTOR_HARVESTER",
        "name": "Harvester Work",
        "name_te": "హార్వెస్టర్ పని",
        "service_category": "tractor_work",
        "default_rate_type": "hourly",
    },
    {
        "code": "TRACTOR_DRONE",
        "name": "Drone Spraying",
        "name_te": "డ్రోన్ పిచికారీ",
        "service_category": "tractor_work",
        "default_rate_type": "hourly",
    },
    # Transport
    {
        "code": "TRANSPORT_BOLERO",
        "name": "Mahindra Bolero Carrying",
        "name_te": "మహీంద్రా బోలేరో రవాణా",
        "service_category": "transport",
        "default_rate_type": "fixed",
    },
    {
        "code": "TRANSPORT_DCM",
        "name": "Eicher DCM Carrying",
        "name_te": "ఐషర్ డీసీఎం రవాణా",
        "service_category": "transport",
        "default_rate_type": "fixed",
    },
    # Fertiliser
    {
        "code": "FERT_UREA",
        "name": "Urea Supply",
        "name_te": "యూరియా సరఫరా",
        "service_category": "fertiliser",
        "default_rate_type": "fixed",
    },
    {
        "code": "FERT_DAP",
        "name": "DAP Supply",
        "name_te": "డీఏపీ సరఫరా",
        "service_category": "fertiliser",
        "default_rate_type": "fixed",
    },
    {
        "code": "FERT_OTHER",
        "name": "Other Fertiliser",
        "name_te": "ఇతర ఎరువులు",
        "service_category": "fertiliser",
        "default_rate_type": "fixed",
    },
    # Seeds
    {
        "code": "SEED_PADDY",
        "name": "Paddy Seeds",
        "name_te": "వరి విత్తనాలు",
        "service_category": "seeds",
        "default_rate_type": "fixed",
    },
    {
        "code": "SEED_CORN",
        "name": "Corn Seeds",
        "name_te": "మొక్కజొన్న విత్తనాలు",
        "service_category": "seeds",
        "default_rate_type": "fixed",
    },
    {
        "code": "SEED_OTHER",
        "name": "Other Seeds",
        "name_te": "ఇతర విత్తనాలు",
        "service_category": "seeds",
        "default_rate_type": "fixed",
    },
    # Agri-finance
    {
        "code": "AGRI_FINANCE_LOAN",
        "name": "Agri Finance Loan",
        "name_te": "వ్యవసాయ రుణం",
        "service_category": "agri_finance",
        "default_rate_type": "fixed",
    },
    {
        "code": "AGRI_FINANCE_ADVANCE",
        "name": "Agri Finance Advance",
        "name_te": "వ్యవసాయ అడ్వాన్స్",
        "service_category": "agri_finance",
        "default_rate_type": "fixed",
    },
    # Vehicle ops
    {
        "code": "VEHICLE_REPAIR",
        "name": "Vehicle Repair",
        "name_te": "వాహన మరమ్మతు",
        "service_category": "vehicle_ops",
        "default_rate_type": "fixed",
    },
    {
        "code": "VEHICLE_MAINTENANCE",
        "name": "Vehicle Maintenance",
        "name_te": "వాహన నిర్వహణ",
        "service_category": "vehicle_ops",
        "default_rate_type": "fixed",
    },
    {
        "code": "VEHICLE_CLEANING",
        "name": "Vehicle Cleaning",
        "name_te": "వాహన శుభ్రత",
        "service_category": "vehicle_ops",
        "default_rate_type": "fixed",
    },
    # Godown
    {
        "code": "GODOWN_REPAIR",
        "name": "Godown Repair",
        "name_te": "గోడౌన్ మరమ్మతు",
        "service_category": "godown",
        "default_rate_type": "fixed",
    },
    {
        "code": "GODOWN_PURCHASE",
        "name": "Godown Purchase",
        "name_te": "గోడౌన్ కొనుగోలు",
        "service_category": "godown",
        "default_rate_type": "fixed",
    },
    {
        "code": "GODOWN_CLEANING",
        "name": "Godown Cleaning",
        "name_te": "గోడౌన్ శుభ్రత",
        "service_category": "godown",
        "default_rate_type": "fixed",
    },
]
